### PR TITLE
Harmonize the use of reco::TrackBase::TrackAlgorithm enumeration

### PR DESCRIPTION
**Pull request by makortel**
This PR is a preparatory step towards adding enumerators for Phase1-specific iterations to reco::TrackBase::TrackAlgorithm enumeration. For now, the he intent is to preserve the current behaviour.

Comparing reco::TrackBase::TrackAlgorithm enumerators with equality on "magic numbers" and with less/greater than on anything are, in general, considered harmful. This PR replaces these numbers with enumerators, and the less/greater than comparisons with switches with explicit enumerators. In RecoTracker/FinalTrackSelectors/plugins/*ListMerger.cc an ordering between enumerators is genuinely needed, for them a separate array mapping the enumerators for their order "priority" is added (constructed compile-time).

Tested in CMSSW_7_5_X_2015-03-09-0200 with 9k events of RelValTTbar. No regression expected, none observed.

@rovere @VinInn @boudoul @venturia

### CODE REVIEW

**Comment by makortel** [comment]
We considered using the C++11 strongly-typed enum for the TrackAlgorithm, but decided to postpone that until there is no need to protect it with #ifdef in the header.

File: ElectroWeakAnalysis/Utilities/python/distTCMET_cfi.py [hunk 1]
```diff
@@ -42,7 +42,7 @@
     chi2_tight_max = cms.double(3.0),
     nhits_tight_min = cms.double(11),
     ptErr_tight_max = cms.double(0.1),
-    maxTrackAlgo = cms.int32(8)
+    trackAlgos = cms.vstring("undefAlgorithm", "ctf", "rs", "cosmics", "initialStep", "lowPtTripletStep", "pixelPairStep", "detachedTripletStep"),
```
Comment (makortel) [inline ElectroWeakAnalysis/Utilities/python/distTCMET_cfi.py]: I'm not perfectly happy with ending up to two parameters with similar names (`trackAlgos` and `track_algos`) and different semantics, but I think this issue kind-of existed already before.

**Comment by cvuosalo** [comment]
@cmsbuild please test

**Comment by cvuosalo** [comment]
Probably for a future PR: Static analyzer has reported problems that might need fixing.

MuonTCMETValueMapProducer.cc, line 290, unused assignment to variable d0 can be deleted.
PhotonValidator.cc, lines 3104-3110, variables bcPhi1 and bcPhi2 are set but never used.

**Comment by cvuosalo** [comment]
+1

For #8315 d63c3c2b4ec6c14fdbd4065ecc0ce25918ee8363

This PR replaces numeric constants representing algorithm codes with enumerators. It is a technical code improvement, not a change in reconstruction.

The code changes are beneficial, though a bit complex (recursive templates!). The Jenkins results are OK, and short matrix tests against baseline CMSSW_7_5_X_2015-03-15 show no significant differences, as expected.

**Comment by deguio** [comment]
+1

**Comment by monttj** [comment]
+1

**Comment by davidlange6** [comment]
+1

**Pull request merged**